src/ccache.cpp

```diff
@@ -871,124 +861,102 @@ process_preprocessed_file(Context& ctx,
-        result = true;
-      } else {
-        relpath = token;
-      }
-      if (token != buf) { // This is a dependency file.
-        fputc(' ', tmpf);
-      }
-      fputs(relpath, tmpf);
-      if (relpath != token) {
-        free(relpath);
-      }
-      token = strtok_r(nullptr, " \t", &saveptr);
+  std::string adjusted_file_content;
+  adjusted_file_content.reserve(file_content.size());
+
+  bool rewritten = false;
+
+  for (string_view token : Util::split_into_views(file_content, " \t\r\n")) {
+    if (Util::is_absolute_path(token)
+        && token.starts_with(ctx.config.base_dir())) {
+      adjusted_file_content.append(make_relative_path(ctx, token));
+      rewritten = true;
+    } else {
+      adjusted_file_content.append(token.begin(), token.end());
     }
+    adjusted_file_content.push_back(' ');
   }
 
-  if (ferror(f)) {
-    cc_log("Error reading dependency file: %s, skip relative path usage",
-           depfile);
-    result = false;
-    goto out;
+  if (!rewritten) {
+    cc_log(
+      "No paths in dependency file %s made relative, skip relative path usage",
+      output_dep.c_str());
+    return;
   }
-  if (ferror(tmpf)) {
+
+  std::string tmp_file = fmt::format("{}.tmp{}", output_dep, tmp_string());
+
+  try {
+    Util::write_file(tmp_file, adjusted_file_content);
+  } catch (const Error& e) {
     cc_log(
-      "Error writing temporary dependency file: %s, skip relative path"
+      "Error writing temporary dependency file %s (%s), skip relative path"
       " usage",
-      tmp_file);
-    result = false;
-    goto out;
+      tmp_file.c_str(),
+      e.what());
+    x_unlink(tmp_file.c_str());
+    return;
   }
 
-out:
-  fclose(tmpf);
-  fclose(f);
-  if (result) {
-    if (x_rename(tmp_file, depfile) != 0) {
-      cc_log(
-        "Error renaming dependency file: %s -> %s (%s), skip relative"
-        " path usage",
-        tmp_file,
-        depfile,
-        strerror(errno));
-      result = false;
-    } else {
-      cc_log("Renamed dependency file: %s -> %s", tmp_file, depfile);
-    }
-  }
-  if (!result) {
-    cc_log("Removing temporary dependency file: %s", tmp_file);
-    tmp_unlink(tmp_file);
+  if (x_rename(tmp_file.c_str(), output_dep.c_str()) != 0) {
+    cc_log(
+      "Error renaming dependency file: %s -> %s (%s), skip relative path usage",
+      tmp_file.c_str(),
+      output_dep.c_str(),
+      strerror(errno));
+    x_unlink(tmp_file.c_str());
+  } else {
+    cc_log("Renamed dependency file: %s -> %s",
+           tmp_file.c_str(),
+           output_dep.c_str());
   }
-  free(tmp_file);
 }
 
 // Extract the used includes from the dependency file. Note that we cannot
 // distinguish system headers from other includes here.
 static struct digest*
 result_name_from_depfile(Context& ctx, struct hash* hash)
 {
-  FILE* f = fopen(ctx.args_info.output_dep.c_str(), "r");
-  if (!f) {
+  std::string file_content;
+  try {
+    file_content = Util::read_file(ctx.args_info.output_dep);
```

Comment:
This pattern is kind of strange. **After** this pull request we should consider other options. Fine for now (topic is too huge to be delayed by this).